ipa-core/src/secret_sharing/vector/transpose.rs

```diff
@@ -0,0 +1,800 @@
+            y3[i].to_le_bytes();
+    }
+    dst
+}
+
+// Degenerate transposes.
+
+impl TransposeFrom<AdditiveShare<BA256, 1>> for Vec<AdditiveShare<BA256>> {
+    fn transpose_from(&mut self, src: AdditiveShare<BA256, 1>) {
+        *self = vec![src];
+    }
+}
+
+impl TransposeFrom<Vec<StdArray<Boolean, 1>>> for Vec<BA256> {
+    fn transpose_from(&mut self, src: Vec<StdArray<Boolean, 1>>) {
+        *self = vec![src.iter().map(Boolean::from_array).collect::<BA256>()];
+    }
+}
+
+/// Perform a larger transpose using an 16x16 kernel.
+///
+/// Matrix height and width must be multiples of 16.
+#[inline]
+fn do_transpose_16<SF: Fn(usize, usize) -> [u8; 32], DF: FnMut(usize, usize, [u8; 32])>(
+    rows_div16: usize,
+    cols_div16: usize,
+    read_src: SF,
+    mut write_dst: DF,
+) {
+    for i in 0..rows_div16 {
+        for j in 0..cols_div16 {
+            let m = read_src(i, j);
+            let m_t = transpose_16x16(&m);
+            write_dst(j, i, m_t);
+        }
+    }
+}
+
+/// Implement a transpose of a MxN bit matrix represented as `[BA{N}; {M}]` into a NxM bit matrix
+/// represented as `[BA{M}; {N}]`.
+///
+/// The invocation looks like `impl_transpose_ba_to_ba!(BA<m>, BA<n>, <m>, <n>)`. e.g. for MxN = 16x64,
+/// `impl_transpose_ba_to_ba!(BA16, BA64, 16, 64)`. Or to put it differently, write the MxN dimensions
+/// twice, first with BA in front, and then without.
+macro_rules! impl_transpose_ba_to_ba {
+    ($dst_row:ty, $src_row:ty, $src_rows:expr, $src_cols:expr, $test_fn:ident) => {
+        impl TransposeFrom<&[$src_row; $src_rows]> for [$dst_row; $src_cols] {
+            fn transpose_from(&mut self, src: &[$src_row; $src_rows]) {
+                do_transpose_16(
+                    $src_rows / 16,
+                    $src_cols / 16,
+                    |i, j| {
+                        let mut d = [0u8; 32];
+                        for k in 0..16 {
+                            d[2 * k..2 * (k + 1)].copy_from_slice(
+                                &src[16 * i + k].as_raw_slice()[2 * j..2 * (j + 1)],
+                            );
+                        }
+                        d
+                    },
+                    |i, j, d| {
+                        for k in 0..16 {
+                            self[16 * i + k].as_raw_mut_slice()[2 * j..2 * (j + 1)]
+                                .copy_from_slice(&d[2 * k..2 * (k + 1)]);
+                        }
+                    },
+                );
+            }
+        }
+
+        #[cfg(all(test, unit_test))]
+        #[test]
+        fn $test_fn() {
+            tests::test_transpose_ba_to_ba::<$dst_row, $src_row, $src_rows, $src_cols>();
+        }
+
+        impl TransposeFrom<&BitDecomposed<$src_row>> for Vec<$dst_row> {
+            fn transpose_from(&mut self, src: &BitDecomposed<$src_row>) {
+                self.resize($src_cols, <$dst_row>::ZERO);
+                let src = <&[$src_row; $src_rows]>::try_from(&**src).unwrap();
```

Comment:
these `unwrap` calls - can they ever fail? if so, should we make `transpose_from` fallible or use `expect` if invariants should always hold?

Reply:
Some of them can fail. I am adding an `Error` type to the transpose trait to handle those cases.

The unwraps that are left should only be for infallible `array::try_from`. It might be nice to annotate these cases (and similar conversions like `<usize as TryFrom<u32>>`) somehow, but it feels like something more structured than `expect("string")` is appropriate. Maybe more unwrap traits like `UnwrapInfallible`? (These would not be quite the same though, since `UnwrapInfallible` can truly never fail.)